File: Reqnroll/Shared/WIMP.App/Services/MenuService.cs
/*
 * NOTE: This application ("WIMP - Where Is My Pizza") is provided solely to
 * demonstrate the Behavior-Driven Development scenario automation patterns
 * described in the book "Effective Behavior-Driven Development" by
 * Gaspar Nagy and Seb Rose.
 *
 * It is NOT a complete or production-ready implementation. It deliberately
 * uses shortcuts and simplifications (e.g. authentication, data storage,
 * error handling, security) that are NOT suitable for a real application.
 * Do not use this code as a basis for production software.
 */

/*
 * NOTE: This application ("WIMP - Where Is My Pizza") is provided solely to
 * demonstrate the Behavior-Driven Development scenario automation patterns
 * described in the book "Effective Behavior-Driven Development" by
 * Gaspar Nagy and Seb Rose.
 *
 * It is NOT a complete or production-ready implementation. It deliberately
 * uses shortcuts and simplifications (e.g. authentication, data storage,
 * error handling, security) that are NOT suitable for a real application.
 * Do not use this code as a basis for production software.
 */

using WIMP.App.Data;
using WIMP.App.Models;

namespace WIMP.App.Services;

public class MenuService(IDataRepository repository)
{
    public IReadOnlyCollection<OfferedMenuItem> LoadMenu(string promotionName = "", MenuFilter? filter = null)
    {
        var offeredItems = repository.GetMenuItems()
            .OrderBy(mi => mi.Name)
            .Select(mi => new OfferedMenuItem
            {
                Name = mi.Name,
                OriginalPrice = mi.Price,
                Price = mi.Price,
                Calories = mi.Calories,
                Vegetarian = mi.Vegetarian
            })
            .ToList();
        if (promotionName == "Friday $1 off")
        {
            offeredItems.ForEach(mi =>
            {
                if (mi.OriginalPrice > 8.00m)
                {
                    mi.Price = mi.OriginalPrice - 1.00m;
                }
            });
        }
        else if (promotionName == "Veggie week")
        {
            offeredItems.ForEach(mi =>
            {
                if (mi.Vegetarian)
                {
                    mi.Price = 6.00m;
                }
            });
        }

        return filter != null ? FilterMenuItems(offeredItems, filter) : offeredItems;
    }

    private IReadOnlyCollection<OfferedMenuItem> FilterMenuItems(IEnumerable<OfferedMenuItem> menuItems, MenuFilter filter)
    {
        if (filter.MinPrice != null)
        {
            menuItems = menuItems.Where(mi => mi.Price >= filter.MinPrice);
        }
        if (filter.MaxPrice != null)
        {
            menuItems = menuItems.Where(mi => mi.Price <= filter.MaxPrice);
        }
        if (filter.MaxCalories != null)
        {
            menuItems = menuItems.Where(mi => mi.Calories <= filter.MaxCalories);
        }

        return menuItems.ToArray();
    }

    public ServiceResult<MenuItem> AddMenuItem(MenuItem menuItem)
    {
        repository.InsertMenuItem(menuItem);
        return ServiceResult<MenuItem>.Success(menuItem);
    }
}

File: CucumberJVM/Shared/WIMP.App/java/com/wimp/app/services/MenuService.java
/*
 * NOTE: This application ("WIMP - Where Is My Pizza") is provided solely to
 * demonstrate the Behavior-Driven Development scenario automation patterns
 * described in the book "Effective Behavior-Driven Development" by
 * Gaspar Nagy and Seb Rose.
 *
 * It is NOT a complete or production-ready implementation. It deliberately
 * uses shortcuts and simplifications (e.g. authentication, data storage,
 * error handling, security) that are NOT suitable for a real application.
 * Do not use this code as a basis for production software.
 */

package com.wimp.app.services;

import com.wimp.app.data.DataRepository;
import com.wimp.app.models.MenuFilter;
import com.wimp.app.models.MenuItem;
import com.wimp.app.models.OfferedMenuItem;
import com.wimp.app.models.ServiceResult;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

@Service
public class MenuService {
    private final DataRepository repository;

    public MenuService(DataRepository repository) {
        this.repository = repository;
    }

    public List<OfferedMenuItem> loadMenu() {
        return loadMenu("", null);
    }

    public List<OfferedMenuItem> loadMenu(String promotionName, MenuFilter filter) {
        List<OfferedMenuItem> offeredItems = repository.getMenuItems().stream()
            .sorted(Comparator.comparing(MenuItem::getName))
            .map(this::toOfferedMenuItem)
            .toList();

        if ("Friday $1 off".equals(promotionName)) {
            offeredItems.forEach(item -> {
                if (item.getOriginalPrice().compareTo(BigDecimal.valueOf(8.00)) > 0) {
                    item.setPrice(item.getOriginalPrice().subtract(BigDecimal.ONE));
                }
            });
        } else if ("Veggie week".equals(promotionName)) {
            offeredItems.forEach(item -> {
                if (item.isVegetarian()) {
                    item.setPrice(BigDecimal.valueOf(6.00));
                }
            });
        }

        if (filter == null) {
            return offeredItems;
        }

        List<OfferedMenuItem> filtered = new ArrayList<>(offeredItems);
        if (filter.getMinPrice() != null) {
            filtered.removeIf(item -> item.getPrice().compareTo(filter.getMinPrice()) < 0);
        }
        if (filter.getMaxPrice() != null) {
            filtered.removeIf(item -> item.getPrice().compareTo(filter.getMaxPrice()) > 0);
        }
        if (filter.getMaxCalories() != null) {
            filtered.removeIf(item -> item.getCalories() > filter.getMaxCalories());
        }
        return filtered;
    }

    public ServiceResult<MenuItem> addMenuItem(MenuItem menuItem) {
        repository.insertMenuItem(menuItem);
        return ServiceResult.success(menuItem);
    }

    private OfferedMenuItem toOfferedMenuItem(MenuItem menuItem) {
        OfferedMenuItem offered = new OfferedMenuItem();
        offered.setName(menuItem.getName());
        offered.setPrice(menuItem.getPrice());
        offered.setCalories(menuItem.getCalories());
        offered.setVegetarian(menuItem.isVegetarian());
        offered.setOriginalPrice(menuItem.getPrice());
        return offered;
    }
}
